Add wedge-section torus InterpolationTarget

fmahebert opened the pull request:
## Proposed changes

Adds an InterpolationTarget `WedgeSectionTorus` that computes the (r, theta, phi) grid for the accretion disk diagnostics.

Commit order:
1. Reorganize InterpTargetLineSegment test to facilitate factoring
2. Factor InterpolationTarget test work
3. Add InterpolationTargetWedgeSectionTorus

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

fmahebert committed:
Reorganize InterpTargetLineSegment test to facilitate factoring

fmahebert committed:
Add InterpolationTargetWedgeSectionTorus

fmahebert committed:
Factor InterpolationTarget test work

fmahebert commented:
This PR is failing to synchronize: I've pushed a minor fix (moving one change from one commit to another) to my fork, but the fix has not propagated to the PR. As I type this, the commits on my fork are different from those in the PR.

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,220 @@
+
+namespace intrp {
+
+namespace OptionHolders {
```
This namespace should be documented so it shows up and we can find `WedgeSectionTorus`. (That may not be your fault, I just checked the docs and noticed)

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,220 @@
+      tmpl::list<MinRadius, MaxRadius, NumberOfRadialPoints, ThetaHalfAngle,
+                 NumberOfThetaPoints, NumberOfPhiPoints>;
+  static constexpr OptionString help = {
+      "A torus extending from MinRadius to MaxRadius in r, ThetaHalfAngle"
```
missing space between `ThetaHalfAngle` and `above`. Could you please add that?

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,220 @@
+      "A torus extending from MinRadius to MaxRadius in r, ThetaHalfAngle"
+      "above and below the equator in theta, and 2pi in phi."};
+
+  WedgeSectionTorus(const double min_radius_in, const double max_radius_in,
```
Let's not inline the constructor. Could you move it outside the class definition?

fmahebert replied:
Keep it in the header, or in a new `.cpp` file? Shall I also fix this in the previous `LineSegment` target, from which I copied the overall design?

fmahebert replied:
@nilsdeppe ping. Do you have a preference on where I move the implementation to?

nilsdeppe replied:
I think a cpp file is fine (unless it's a template, then just do header). If you put it in the header and the class isn't a template you'll be back to needing to mark it `inline` :)

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,220 @@
+  ~WedgeSectionTorus() = default;
+
+  // clang-tidy non-const reference pointer.
+  void pup(PUP::er& p) noexcept {  // NOLINT
```
Let's also not inline the `pup` method.

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,220 @@
+  struct MinRadius {
+    using type = double;
+    static constexpr OptionString help = {"Inner radius of torus"};
+    static type lower_bound() { return 0.0; }
```
I'm not sure how uniformly we have been doing this, but I believe the `bound` functions can all be `noexcept`

fmahebert replied:
Sure, I'll fix this one. More generally in the codebase, very few of these `bound` functions are marked `noexcept`: I'm also happy to fix this if you assign an issue to me.

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,220 @@
+
+    // Compute locations of constant r/theta/phi surfaces
+    const size_t num_radial = options.number_of_radial_points;
+    const DataVector radii_1d = [&num_radial, &options]() {
```
`noexcept`

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,220 @@
+      return result;
+    }();
+    const size_t num_theta = options.number_of_theta_points;
+    const DataVector thetas_1d = [&num_theta, &options]() {
```
`noexcept`

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,220 @@
+      return result;
+    }();
+    const size_t num_phi = options.number_of_phi_points;
+    const DataVector phis_1d = [&num_phi]() {
```
`noexcept`

nilsdeppe reviewed (changes requested):
Because of the conflicts I ignored the first and last commit since I have no clue how/what will change because of it. Could you please rebase this at your earliest convenience and ping me to review those commits? Thanks! :)

fmahebert commented:
@nilsdeppe: conflict resolved

fmahebert committed:
Improve documentation of intrp and intrp::OptionHolders namespaces

wthrowe commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,220 @@
+  struct NumberOfPhiPoints {
+    using type = size_t;
+    static constexpr OptionString help = {
+        "Number of phi points, including endpoints"};
```
Endpoints in phi?

fmahebert replied:
Whoops

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Interpolation/InterpolationTargetTestHelpers.hpp:
```diff
@@ -0,0 +1,263 @@
+// Distributed under the MIT License.
```
Maybe just name the file `TargetTestHelpers.hpp` since it's already in `Interpolation`?

fmahebert replied:
I think it's more natural to keep this filename for now (so that it is consistent with everything else in the directory), and go back to give it a cleaner name as part of fixing issue #1095. If we're agreed, I can add a note to that issue reminding us to also reorganize the tests and this helper as well.

nilsdeppe reviewed (commented):
Couple minor things, overall I'm happy with this. @markscheel your input would be greatly appreciated here :)

moxcodes commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolationTargetLineSegment.cpp:
```diff
@@ -67,11 +67,9 @@ struct mock_interpolation_target {
-                                                 Frame::Inertial>>>;
+  using component_being_mocked = void;  // not needed.
+  using const_global_cache_tag_list = Parallel::get_const_global_cache_tags<
+      tmpl::list<typename Metavariables::compute_target_points>>;
```
for `compute_target_points` - keep it in the `InterpolationTargetA` struct - the interpolation targets will each have its own `compute_target_points`, so needs the scope. - @markscheel

moxcodes commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolationTargetLineSegment.cpp:
```diff
@@ -142,10 +140,11 @@ struct MockMetavariables {
+    using type = ::intrp::Actions::LineSegment<InterpolationTargetA, 3,
+                                               Frame::Inertial>::options_type;
   };
+  using compute_target_points =
```
see above

moxcodes commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolationTargetLineSegment.cpp:
```diff
@@ -214,74 +215,91 @@ SPECTRE_TEST_CASE("Unit.NumericalAlgorithms.InterpolationTarget.LineSegment",
+  runner.template simple_action<
+      mock_interpolation_target<metavars,
+                                typename metavars::InterpolationTargetA>,
+      typename metavars::compute_target_points>(0, temporal_id);
```
as above

moxcodes commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolationTargetLineSegment.cpp:
```diff
@@ -214,74 +215,91 @@ SPECTRE_TEST_CASE("Unit.NumericalAlgorithms.InterpolationTarget.LineSegment",
+  const auto domain_creator =
+      DomainCreators::Shell<Frame::Inertial>(0.9, 4.9, 1, {{5, 5}}, false);
+
+  const auto expected_block_coord_holders = [&domain_creator]() {
```
noexcept

moxcodes commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,220 @@
+/// \brief A solid torus of points, useful, e.g., when measuring data from an
+/// accretion disc.
+///
+/// The torus' cross section (e.g., a cut at \f$\phi=0\f$) is a wedge-like shape
```
we decided `torus's`

moxcodes commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,220 @@
+/// \note At least two grid points are required per dimension.
+///
+/// \note Input coordinates (radii, angles) are interpreted in the frame given
+/// by the `Frame` template parameter of the `Actions::WedgeSectionTorus`.
```
add same note to `LineSegment`

moxcodes commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,220 @@
+    const DataVector phis_1d = [&num_phi]() {
+      DataVector result(num_phi);
+      for (size_t phi = 0; phi < num_phi; ++phi) {
+        result[phi] = 2.0 * M_PI * phi / (num_phi - 1.0);
```
have a duplicate point at 0 and 2pi (so divide by just `num_phi`)

moxcodes commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,220 @@
+      return result;
+    }();
+
+    // Take tensor product to get full 3D r/theta/phi points
```
document loop order

moxcodes commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolationTargetWedgeSectionTorus.cpp:
```diff
@@ -0,0 +1,83 @@
+        ::intrp::Actions::WedgeSectionTorus<InterpolationTargetA,
+                                            Frame::Inertial>::options_type;
+  };
+  using compute_target_points =
```
move into struct as previous

moxcodes commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolationTargetWedgeSectionTorus.cpp:
```diff
@@ -0,0 +1,83 @@
+        const double theta =
+            (M_PI / 8.0) * (-1.0 + 2.0 * t / (num_theta - 1.0));
+        for (size_t p = 0; p < num_phi; ++p) {
+          const double phi = 2.0 * M_PI * p / (num_phi - 1.0);
```
`num_phi - 1.0` -> `num_phi`

moxcodes reviewed (changes requested):
Group code review with @fmahebert, @markscheel, @mgiesler.

We also discussed that the theta and r points should probably be Gauss-Lobatto for applications like integrals, but uniform may be best for visualization. Perhaps an option should be supplied. @fmahebert has volunteered to add this option. @nilsdeppe @kidder @wthrowe (and anyone else) thoughts?

fmahebert committed:
Improve documentation of LineSegment InterpolationTarget

nilsdeppe commented:
@fmahebert this needs a rebase

fmahebert commented:
Summary of discussion on telecon: Yes, we probably want r, theta points arranged in Gauss-Lobatto way. We may even want to subdivide these directions into pseudo-elements (pseudo because do not want two points (one point per side) at boundaries for the interpolation problem). Smaller elements would reduce the need to use very high order polynomials when representing the solution. For the phi direction, probably still want to keep the full 2pi Fourier basis.

fmahebert commented:
Updated, rebased, and squashed.

I decided not to implement the pseudo-elements we discussed last week, because I'm not sure of the best way to do that right now. Instead I have generalized the interface to give the torus arbitrary bounds theta_min and theta_max — this makes it possible to "manually" subdivide the torus into different r, theta rings if necessary.

Edit for clarity: in the previous sentence, I meant that the total set of points can be broken up into different InterpolationTargets, each of which takes a subset of points in r and/or theta.

We can always come back with a better solution once we have a better understanding of what we need.

markscheel commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetLineSegment.hpp:
```diff
@@ -19,6 +19,9 @@ namespace intrp {
 /// containing `NumberOfPoints` uniformly-spaced points including the endpoints.
+///
+/// \note Input coordinates are interpreted in the frame given by the `Frame`
+/// by the `Frame` template parameter of the `Actions::LineSegment`.
```
This comment will need to change when #1132 is merged.  After that change, the coordinates are interpreted in the frame given by Metavariables::domain_frame.

markscheel commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,278 @@
+/// The `target_points` form a 3D mesh ordered with \f$r\f$ varying fastest,
+/// then \f$\theta\f$, and finally \f$\phi\f$ varying slowest.
+///
+/// \note At least two grid points are required per dimension.
```
I think we can require only one phi point.

markscheel commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,278 @@
+  struct NumberOfPhiPoints {
+    using type = size_t;
+    static constexpr OptionString help = {"Number of phi points"};
+    static type lower_bound() noexcept { return 2; }
```
I think one phi point would work.

fmahebert commented:
Updated to allow single point in phi and squashed directly. Rebased on master to address conflicts from #1131. Need to decide if this gets merged before or after #1132 (because of conflicts and the documentation change @markscheel indicated in his last review).

fmahebert commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,276 @@
+
+  using options =
+      tmpl::list<MinRadius, MaxRadius, MinTheta, MaxTheta, NumberOfRadialPoints,
+                 NumberOfThetaPoints, NumberOfPhiPoints>;
```
RIP

moxcodes commented on tests/Unit/NumericalAlgorithms/Interpolation/InterpolationTargetTestHelpers.hpp:
```diff
@@ -0,0 +1,265 @@
+            const gsl::not_null<
+                db::item_type<intrp::Tags::InterpolatedVarsHolders<
+                    Metavariables, VolumeDim>>*>
+                vars_holders) {
```
noexcept

moxcodes commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolationTargetWedgeSectionTorus.cpp:
```diff
@@ -0,0 +1,134 @@
+  enum class Phase { Initialize, Exit };
+};
+
+void test_r_theta_lgl() {
```
noexcept

moxcodes commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolationTargetWedgeSectionTorus.cpp:
```diff
@@ -0,0 +1,134 @@
+      domain_creator, wedge_section_torus_opts, expected_block_coord_holders);
+}
+
+void test_r_theta_uniform() {
```
noexcept

wthrowe commented on src/NumericalAlgorithms/Interpolation/CMakeLists.txt:
```diff
@@ -6,6 +6,7 @@ set(LIBRARY Interpolation)
 set(LIBRARY_SOURCES
   BarycentricRational.cpp
   IrregularInterpolant.cpp
+  InterpolationTargetWedgeSectionTorus.cpp
```
Alphabetize.

wthrowe commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -0,0 +1,276 @@
+/// \note Input coordinates (radii, angles) are interpreted in the frame given
+/// by the `Frame` template parameter of the `Actions::WedgeSectionTorus`.
+struct WedgeSectionTorus {
+  struct MinRadius {
```
Test creation.

moxcodes reviewed (commented):
group code review with @markscheel @fmahebert @geoffrey4444

wthrowe commented:
Looks good to me.  I think I'm the only one with nontrivial requests open, so I'm happy if you squash.

fmahebert committed:
Test creation of LineSegment opts
